tests/e2e/e2e_test.go

```diff
@@ -256,29 +334,406 @@ func (s *IntegrationTestSuite) TestConcentratedLiquidity() {
+	deltaSqrtPriceAfterNextInitializedTick := sqrtPriceAfterNextInitializedTick.Sub(sqrtPriceAtNextInitializedTick)
+	deltaSqrtPriceAtNextInitializedTick := sqrtPriceAtNextInitializedTick.Sub(sqrtPriceBeforeSwap)
+
+	// Calculate the amount of osmo required to:
+	// * amountInToGetToTickAfterInitialized - move price from next initialized tick (400) to destination tick (400 + tickOffset)
+	// * amountInToGetToNextInitTick - move price from current tick to next initialized tick
+	// Formula is as follows:
+	// Δy = L * Δ(sqrtPrice)
+	amountInToGetToTickAfterInitialized := deltaSqrtPriceAfterNextInitializedTick.Mul(liquidityBeforeSwap.Sub(positionsAddress1[0].Liquidity))
+	amountInToGetToNextInitTick := deltaSqrtPriceAtNextInitializedTick.Mul(liquidityBeforeSwap)
+
 	var (
+		// Swap parameters
+
+		// uosmoInDec_Swap2_NoFee is calculated such that swapping this amount (not considering fee) moves the price over the next initialized tick
+		uosmoInDec_Swap2_NoFee = amountInToGetToNextInitTick.Add(amountInToGetToTickAfterInitialized)
+		uosmoInDec_Swap2       = uosmoInDec_Swap2_NoFee.Quo(sdk.OneDec().Sub(swapFeeDec)).TruncateDec() // account for swap fee of 1%
+		uosmoIn_Swap2          = fmt.Sprintf("%suosmo", uosmoInDec_Swap2.String())
+
+		feeGrowthGlobal_Swap1 = feeGrowthGlobal.Clone()
+	)
+	// Perform a swap
+	node.SwapExactAmountIn(uosmoIn_Swap2, outMinAmt, fmt.Sprintf("%d", poolID), denom0, initialization.ValidatorWalletName)
+
+	// Calculate the amount of liquidity of the position that was kicked out during swap (address1 position1)
+	liquidityOfKickedOutPosition := positionsAddress1[0].Liquidity
+
+	// Update pool and track pool's liquidity
+	concentratedPool = updatedPool(poolID)
+
+	liquidityAfterSwap = concentratedPool.GetLiquidity()
+
+	// Assert that net liquidity of kicked out position was successfully removed from current pool's liquidity
+	s.Require().Equal(liquidityBeforeSwap.Sub(liquidityOfKickedOutPosition), liquidityAfterSwap)
+
+	// Collect fees: Swap 2
+
+	// Calculate fee charges per each step
+
+	// Step1: amountIn is uosmo tokens that are swapped + uosmo tokens that are paid for fee
+	// hasReachedTarget in SwapStep is true, hence, to find fees, calculate:
+	// feeCharge = amountIn * swapFee / (1 - swapFee)
+	feeCharge_Swap2_Step1 := amountInToGetToNextInitTick.Mul(swapFeeDec).Quo(sdk.OneDec().Sub(swapFeeDec))
+
+	// Step2: hasReachedTarget in SwapStep is false (nextTick is 1200), hence, to find fees, calculate:
+	// feeCharge = amountRemaining - amountOne
+	amountRemainingAfterStep1 := uosmoInDec_Swap2.Sub(amountInToGetToNextInitTick).Sub(feeCharge_Swap2_Step1)
+	feeCharge_Swap2_Step2 := amountRemainingAfterStep1.Sub(amountInToGetToTickAfterInitialized)
+
+	// per unit of virtual liquidity
+	feeCharge_Swap2_Step1.QuoMut(liquidityBeforeSwap)
+	feeCharge_Swap2_Step2.QuoMut(liquidityAfterSwap)
+
+	// Update feeGrowthGlobal
+	feeGrowthGlobal.AddMut(feeCharge_Swap2_Step1.Add(feeCharge_Swap2_Step2))
```

Comment:
what's the reason we manually do addition to fee growth global?

Reply:
I added this variable to track global fee growth after each swap, hence, at every swap I calculate and update this variable.

 Afaik, you cannot get global fee growth global by querying state or via some method on `ConcentratedPoolExtension` (which is called `concentratedPool` in this test), so I did not really see other options as to how I can track this variable other than manually calculating it on each swap